Sources/SwiftNetKit/NetworkService.swift

```diff
@@ -7,24 +7,42 @@
 
 import Foundation
 
-public struct NetworkService {
+public struct NetworkService: NetworkServiceProtocol {
     
     public init() {}
     
-    public func get<T: Decodable>(from url: URL, decodeTo type: T.Type) async throws -> T {
+    func start<Request: RequestProtocol>(_ request: Request) async throws -> Request.ResponseType {
         do {
-            let (data, response) = try await URLSession.shared.data(from: url)
+            var urlRequest = URLRequest(url: request.url)
+            
+            if let parameters = request.parameters {
+                let queryItems = parameters.map { key, value in
+                    URLQueryItem(name: key, value: "\(value)")
+                }
+                var urlComponents = URLComponents(url: request.url, resolvingAgainstBaseURL: false)
+                urlComponents?.queryItems = queryItems
+                urlRequest.url = urlComponents?.url
+            }
+            
+            urlRequest.httpMethod = request.method.stringValue
+            urlRequest.allHTTPHeaderFields = request.headers
+            
+            if let body = request.body {
+                urlRequest.httpBody = body
+            }
+            
+            let (data, response) = try await URLSession.shared.data(for: urlRequest)
```

Comment:
I would also have the URLSession be a variable stored in the network service, and probably defined in the protocol, rather than the shared session.  You can edit the init to allow for different URLSession configs, like `default` or `ephemeral`, and manage the delegate for future caching.